### DISCUSSION TITLE
How to read shell arguments?

### DISCUSSION THREAD

**Discussion by Cologler** [nushell/nushell#13973] [Q&A]
I have tried to use `nu -c ... 1` to print the args `1`.

``` shell
❯ nu -c 'echo $args' 1
Error: nu::parser::variable_not_found

  × Variable not found.
   ╭─[source:1:6]
 1 │ echo $args
   ·      ──┬──
   ·        ╰── variable not found.
   ╰────

❯ nu -c 'echo $0' 1
Error: nu::parser::variable_not_found

  × Variable not found.
   ╭─[source:1:6]
 1 │ echo $0
   ·      ─┬
   ·       ╰── variable not found.
   ╰────

❯ nu -c 'echo $1' 1
Error: nu::parser::variable_not_found

  × Variable not found.
   ╭─[source:1:6]
 1 │ echo $1
   ·      ─┬
   ·       ╰── variable not found.
   ╰────

❯ nu -c 'def main [args: int] { echo $args }' 1
// print nothing

❯ nu -c 'def main [...args] { echo $args }' 1
// print nothing
```

---

I can't find any document for this, even asked the ChatGPT.
```

**Comment by ghost** [comment]
nu does not have bashs concept of `$1`, etc. you will have to define a `main` function

```nu
def main [arg1: string] {
  print $arg1
}
```

[documentation for functions](https://www.nushell.sh/book/custom_commands.html)
[documentation for main](https://www.nushell.sh/book/modules.html#main)

EDIT: you could use stdin if you have `nu -c` (`"foo" | nu --stdin -c "print $in"`)

**Reply by Cologler** [reply]
But my last two examples did not work either.

**Reply by ghost** [reply]
i think it might have something to do with the arg-parsing:

```
Usage:
  > nu {flags} (script file) ...(script args)
```

if no script file is provided it probably does not accept arguments (maybe adding a warning would be a good idea)

**Reply by Cologler** [reply]
Did you mean that `script args` does not work with the `--commands <string>` mode?

**Reply by ghost** [reply]
i suspect it.  
since arguments have to be provided after the script-file (which cant be provided when `-c` is active) they might be unprovidable
